src/Storages/System/StorageSystemZooKeeper.cpp

```diff
@@ -28,6 +31,128 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
+// ZkNodeCache is a trie tree to cache all the zookeeper writes. The purpose of this struct is to avoid creating/setting nodes
+// repeatedly. For example, If we create path /a/b/c/d/e and path /a/b/d/f in the same transaction. We don't want to create
+// their common path "/a/b" twice. This data structure will cache this changes and generates the eventual requests within one pass.
+struct ZkNodeCache
+{
+    using ZkNodeCachePtr = std::shared_ptr<ZkNodeCache>;
+
+    std::unordered_map<String, ZkNodeCachePtr> children;
+    String value;
+    String path;
+    bool exists;
+    bool changed;
+
+    ZkNodeCache() : path("/"), exists(true), changed(false) { }
+    ZkNodeCache(String path_, bool exists_) : path(path_), exists(exists_), changed(false) { }
+
+    void insert(const std::vector<String> & nodes, zkutil::ZooKeeperPtr zookeeper, const String & value_to_set, size_t index)
+    {
+        if (index >= nodes.size())
+        {
+            value = value_to_set;
+            changed = true;
+            return;
+        }
+        const String & child_name = nodes[index];
+        index++;
+        if (!children.contains(child_name))
+        {
+            String sub_path = "/" + boost::algorithm::join(std::vector<String>(nodes.begin(), nodes.begin() + index), "/");
+            bool ch_exist = false;
+            if (exists)
+            {
+                // If this node doesn't exists, neither will its child.
+                ch_exist = zookeeper->exists(sub_path);
+            }
+            children[child_name] = std::make_shared<ZkNodeCache>(sub_path, ch_exist);
+        }
+        children[child_name]->insert(nodes, zookeeper, value_to_set, index);
+    }
+
+    void generateRequests(Coordination::Requests & requests)
+    {
+        // If the node doesn't exists, we should generate create request.
+        // If the node exists, we should generate set request.
+        // This dfs will prove ancestor nodes are processed first.
+        if (!exists)
+        {
+            auto request = std::make_shared<Coordination::CreateRequest>();
+            request->path = path;
+            request->data = value;
+            request->is_ephemeral = false;
+            request->is_sequential = false;
+            requests.push_back(request);
+        }
+        else if (changed)
+        {
+            auto request = std::make_shared<Coordination::SetRequest>();
+            request->path = path;
+            request->data = value;
+            request->version = -1;
+            requests.push_back(request);
+        }
+        for (auto [_, child] : children)
+            child->generateRequests(requests);
```

Comment:
Looks like recursion depth is not limited here and depends on user's input only. Should we call `checkStackSize()` here?

Reply:
Good point. I think the better way is to restrict the input path length, which impacts the stack size. What value do you think is ok?

Reply:
Maybe `PATH_MAX` (or just 4096)? But we need a test that tries to insert a sting like this

Reply:
Good idea